### PR TITLE
Avoid temporary string allocations when serializing CSS

### PR DESCRIPTION
**Pull request by weinig**
#### 23e4ef26e7c26f02676a845473585baf81961a84
<pre>
Avoid temporary string allocations when serializing CSS
<a href="https://bugs.webkit.org/show_bug.cgi?id=276318">https://bugs.webkit.org/show_bug.cgi?id=276318</a>

Reviewed by NOBODY (OOPS!).

When serializing CSS rules and values, the serialization often
requires the concatenation of multiple items. To avoid unnecessary
temporary string construction overloads of the `cssText`/`customCSSText`
functions that take a `StringBuilder` can be used to allow writing
the result directly into the final location.

CSS rule subclasses now must implement the `cssText` method that
takes a `StringBuilder` rather than one that returns a `String`. The
latter is implemented only for the base class, and calls to the
former. Additionally, CSS rule subclasses may implement the
`cssTextWithReplacementURLs() method that takes a `StringBuilder`
if they serialize URLs. Again, the version that returns a `String`
is implemented in the base class by calling the `StringBuilder` one.

CSS value subclasses now must implement a `customCSSText` method that
takes a `StringBuilder`, but, unlike rules, they can also implement
an overload that returns a `String` if the implementation would be
more efficient for callers that only want its value. Any subclass
that doesn&apos;t implement the version returning a `String` will instead
have have their `StringBuilder` version called from the base class.

Additional work is still needed avoid temporary allocations for
shorthand and selector serialization.

* Source/WebCore/css/CSSAnchorValue.cpp:
* Source/WebCore/css/CSSAnchorValue.h:
* Source/WebCore/css/CSSAspectRatioValue.cpp:
* Source/WebCore/css/CSSAspectRatioValue.h:
* Source/WebCore/css/CSSBackgroundRepeatValue.cpp:
* Source/WebCore/css/CSSBackgroundRepeatValue.h:
* Source/WebCore/css/CSSBasicShapes.cpp:
* Source/WebCore/css/CSSBasicShapes.h:
* Source/WebCore/css/CSSBorderImageSliceValue.cpp:
* Source/WebCore/css/CSSBorderImageSliceValue.h:
* Source/WebCore/css/CSSBorderImageWidthValue.cpp:
* Source/WebCore/css/CSSBorderImageWidthValue.h:
* Source/WebCore/css/CSSCanvasValue.cpp:
* Source/WebCore/css/CSSCanvasValue.h:
* Source/WebCore/css/CSSContainerRule.cpp:
* Source/WebCore/css/CSSContainerRule.h:
* Source/WebCore/css/CSSContentDistributionValue.cpp:
* Source/WebCore/css/CSSContentDistributionValue.h:
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
* Source/WebCore/css/CSSCounterStyleDescriptors.h:
* Source/WebCore/css/CSSCounterStyleRule.cpp:
* Source/WebCore/css/CSSCounterStyleRule.h:
* Source/WebCore/css/CSSCounterValue.cpp:
* Source/WebCore/css/CSSCounterValue.h:
* Source/WebCore/css/CSSCrossfadeValue.cpp:
* Source/WebCore/css/CSSCrossfadeValue.h:
* Source/WebCore/css/CSSCursorImageValue.cpp:
* Source/WebCore/css/CSSCursorImageValue.h:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSFilterImageValue.cpp:
* Source/WebCore/css/CSSFilterImageValue.h:
* Source/WebCore/css/CSSFontFaceRule.cpp:
* Source/WebCore/css/CSSFontFaceRule.h:
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
* Source/WebCore/css/CSSFontFaceSrcValue.h:
* Source/WebCore/css/CSSFontFeatureValue.cpp:
* Source/WebCore/css/CSSFontFeatureValue.h:
* Source/WebCore/css/CSSFontFeatureValuesRule.cpp:
* Source/WebCore/css/CSSFontFeatureValuesRule.h:
* Source/WebCore/css/CSSFontPaletteValuesOverrideColorsValue.cpp:
* Source/WebCore/css/CSSFontPaletteValuesOverrideColorsValue.h:
* Source/WebCore/css/CSSFontPaletteValuesRule.cpp:
* Source/WebCore/css/CSSFontPaletteValuesRule.h:
* Source/WebCore/css/CSSFontStyleRangeValue.cpp:
* Source/WebCore/css/CSSFontStyleRangeValue.h:
* Source/WebCore/css/CSSFontStyleWithAngleValue.cpp:
* Source/WebCore/css/CSSFontStyleWithAngleValue.h:
* Source/WebCore/css/CSSFontValue.cpp:
* Source/WebCore/css/CSSFontValue.h:
* Source/WebCore/css/CSSFontVariantAlternatesValue.cpp:
* Source/WebCore/css/CSSFontVariantAlternatesValue.h:
* Source/WebCore/css/CSSFontVariationValue.cpp:
* Source/WebCore/css/CSSFontVariationValue.h:
* Source/WebCore/css/CSSFunctionValue.cpp:
* Source/WebCore/css/CSSFunctionValue.h:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
* Source/WebCore/css/CSSGridAutoRepeatValue.cpp:
* Source/WebCore/css/CSSGridAutoRepeatValue.h:
* Source/WebCore/css/CSSGridIntegerRepeatValue.cpp:
* Source/WebCore/css/CSSGridIntegerRepeatValue.h:
* Source/WebCore/css/CSSGridLineNamesValue.cpp:
* Source/WebCore/css/CSSGridLineNamesValue.h:
* Source/WebCore/css/CSSGridTemplateAreasValue.cpp:
* Source/WebCore/css/CSSGridTemplateAreasValue.h:
* Source/WebCore/css/CSSGroupingRule.cpp:
* Source/WebCore/css/CSSGroupingRule.h:
* Source/WebCore/css/CSSImageSetOptionValue.cpp:
* Source/WebCore/css/CSSImageSetOptionValue.h:
* Source/WebCore/css/CSSImageSetValue.cpp:
* Source/WebCore/css/CSSImageSetValue.h:
* Source/WebCore/css/CSSImageValue.cpp:
* Source/WebCore/css/CSSImageValue.h:
* Source/WebCore/css/CSSImportRule.cpp:
* Source/WebCore/css/CSSImportRule.h:
* Source/WebCore/css/CSSKeyframeRule.cpp:
* Source/WebCore/css/CSSKeyframeRule.h:
* Source/WebCore/css/CSSKeyframesRule.cpp:
* Source/WebCore/css/CSSKeyframesRule.h:
* Source/WebCore/css/CSSLayerBlockRule.cpp:
* Source/WebCore/css/CSSLayerBlockRule.h:
* Source/WebCore/css/CSSLayerStatementRule.cpp:
* Source/WebCore/css/CSSLayerStatementRule.h:
* Source/WebCore/css/CSSLineBoxContainValue.cpp:
* Source/WebCore/css/CSSLineBoxContainValue.h:
* Source/WebCore/css/CSSMarkup.cpp:
* Source/WebCore/css/CSSMarkup.h:
* Source/WebCore/css/CSSMediaRule.cpp:
* Source/WebCore/css/CSSMediaRule.h:
* Source/WebCore/css/CSSNamedImageValue.cpp:
* Source/WebCore/css/CSSNamedImageValue.h:
* Source/WebCore/css/CSSNamespaceRule.cpp:
* Source/WebCore/css/CSSNamespaceRule.h:
* Source/WebCore/css/CSSOffsetRotateValue.cpp:
* Source/WebCore/css/CSSOffsetRotateValue.h:
* Source/WebCore/css/CSSPageRule.cpp:
* Source/WebCore/css/CSSPageRule.h:
* Source/WebCore/css/CSSPaintImageValue.cpp:
* Source/WebCore/css/CSSPaintImageValue.h:
* Source/WebCore/css/CSSPendingSubstitutionValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSPropertyRule.cpp:
* Source/WebCore/css/CSSPropertyRule.h:
* Source/WebCore/css/CSSQuadValue.cpp:
* Source/WebCore/css/CSSQuadValue.h:
* Source/WebCore/css/CSSRayValue.cpp:
* Source/WebCore/css/CSSRayValue.h:
* Source/WebCore/css/CSSRectValue.cpp:
* Source/WebCore/css/CSSRectValue.h:
* Source/WebCore/css/CSSReflectValue.cpp:
* Source/WebCore/css/CSSReflectValue.h:
* Source/WebCore/css/CSSRule.cpp:
* Source/WebCore/css/CSSRule.h:
* Source/WebCore/css/CSSScopeRule.cpp:
* Source/WebCore/css/CSSScopeRule.h:
* Source/WebCore/css/CSSScrollValue.cpp:
* Source/WebCore/css/CSSScrollValue.h:
* Source/WebCore/css/CSSSelector.cpp:
* Source/WebCore/css/CSSShadowValue.cpp:
* Source/WebCore/css/CSSShadowValue.h:
* Source/WebCore/css/CSSStartingStyleRule.cpp:
* Source/WebCore/css/CSSStartingStyleRule.h:
* Source/WebCore/css/CSSStyleRule.cpp:
* Source/WebCore/css/CSSStyleRule.h:
* Source/WebCore/css/CSSStyleSheet.cpp:
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/CSSSubgridValue.cpp:
* Source/WebCore/css/CSSSubgridValue.h:
* Source/WebCore/css/CSSSupportsRule.cpp:
* Source/WebCore/css/CSSSupportsRule.h:
* Source/WebCore/css/CSSTimingFunctionValue.cpp:
* Source/WebCore/css/CSSTimingFunctionValue.h:
* Source/WebCore/css/CSSTransformListValue.h:
* Source/WebCore/css/CSSUnicodeRangeValue.cpp:
* Source/WebCore/css/CSSUnicodeRangeValue.h:
* Source/WebCore/css/CSSUnknownRule.h:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSValueList.cpp:
* Source/WebCore/css/CSSValueList.h:
* Source/WebCore/css/CSSValuePair.cpp:
* Source/WebCore/css/CSSValuePair.h:
* Source/WebCore/css/CSSVariableData.cpp:
* Source/WebCore/css/CSSVariableData.h:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
* Source/WebCore/css/CSSVariableReferenceValue.h:
* Source/WebCore/css/CSSViewValue.cpp:
* Source/WebCore/css/CSSViewValue.h:
* Source/WebCore/css/DOMCSSNamespace.cpp:
* Source/WebCore/css/DeprecatedCSSOMValueList.cpp:
* Source/WebCore/css/Quad.h:
* Source/WebCore/css/Rect.h:
* Source/WebCore/css/ShorthandSerializer.cpp:
* Source/WebCore/css/ShorthandSerializer.h:
* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/StylePropertiesInlines.h:
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.cpp:
* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.h:
* Source/WebCore/css/calc/CSSCalcSymbolNode.cpp:
* Source/WebCore/css/calc/CSSCalcSymbolNode.h:
* Source/WebCore/css/calc/CSSCalcValue.cpp:
* Source/WebCore/css/calc/CSSCalcValue.h:
* Source/WebCore/css/parser/CSSParserToken.cpp:
* Source/WebCore/css/parser/CSSParserTokenRange.cpp:
* Source/WebCore/css/parser/CSSParserTokenRange.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnevaluatedCalc.cpp:
* Source/WebCore/css/query/ContainerQuery.cpp:
* Source/WebCore/css/query/GenericMediaQuerySerialization.cpp:
* Source/WebCore/css/query/MediaQueryParser.cpp:
* Source/WebCore/css/typedom/CSSKeywordValue.cpp:
* Source/WebCore/css/typedom/CSSStyleImageValue.cpp:
* Source/WebCore/css/typedom/CSSStyleValue.cpp:
* Source/WebCore/editing/MarkupAccumulator.cpp:
* Source/WebCore/html/HTMLImageElement.cpp:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
* Source/WebCore/page/PageSerializer.cpp:
* Source/WebCore/platform/graphics/ColorSerialization.cpp:
* Source/WebCore/platform/graphics/ColorSerialization.h:
* Source/WebCore/svg/SVGPathParser.cpp:
* Source/WebCore/svg/SVGPathParser.h:
* Source/WebCore/svg/SVGPathStringBuilder.cpp:
* Source/WebCore/svg/SVGPathStringBuilder.h:
* Source/WebCore/svg/SVGPathUtilities.cpp:
* Source/WebCore/svg/SVGPathUtilities.h:
* Source/WebCore/testing/Internals.cpp:
</pre>
https://github.com/WebKit/WebKit/commit/23e4ef26e7c26f02676a845473585baf81961a84

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57595 "15 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36923 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10070 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61217 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8040 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59723 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44559 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8228 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46637 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-inset-009.html (failure)") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5709 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59625 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34644 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49766 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27503 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31420 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-inset-009.html imported/w3c/web-platform-tests/websockets/basic-auth.any.html?wss (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7059 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7043 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53377 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7332 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62897 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1509 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7423 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-inset-009.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53900 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-shapes/shape-outside/values/shape-outside-inset-009.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1515 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49780 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54008 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1286 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32752 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33838 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34922 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33583 "Built successfully") | | |

### CODE REVIEW

**Comment by weinig** [comment]
This PR shows a bit of a weakness in the StringBuilder API that makes writing straight-line code impossible with recursive functions like cssText(StringBuilder&). I filed https://bugs.webkit.org/show_bug.cgi?id=276286 to investigate ways to improve that.

**Comment by weinig** [comment]
One other observation from making this change is that the function `nameLiteral(CSSValueID)` is called a lot in serialization functions, rather than its peer, `nameLiteralForSerialization(CSSValueID)`. I believe the only difference between the two is if capital letters are used for a few weird value IDs like `optimizeLegibility`. Never-the-less, we should probably be calling the latter everywhere as that is what the spec wants -> https://drafts.csswg.org/cssom/#serialize-a-css-component-value.

**Comment by weinig** [comment]
Ok, should be ready for review now.

File: Source/WebCore/css/CSSAnchorValue.h [hunk 1]
```diff
@@ -40,7 +40,8 @@ class CSSAnchorValue final : public CSSValue {
 
     void collectComputedStyleDependencies(ComputedStyleDependencies&) const;
 
-    String customCSSText() const;
+    void customCSSText(StringBuilder&) const;
```
Comment (darinadler) [inline Source/WebCore/css/CSSAnchorValue.h:43]: I think that we should add the word _append_ to these function names when changing them to append to a `StringBuilder` rather than using overloading. Unless there is a reason overloading has a benefit.

File: Source/WebCore/css/CSSBackgroundRepeatValue.cpp [hunk 2]
```diff
@@ -44,17 +44,27 @@ Ref<CSSBackgroundRepeatValue> CSSBackgroundRepeatValue::create(CSSValueID repeat
...
 }
 
-String CSSBackgroundRepeatValue::customCSSText() const
+template<typename Maker> decltype(auto) CSSBackgroundRepeatValue::serialize(Maker&& maker) const
 {
     // background-repeat/mask-repeat behave a little like a shorthand, but `repeat no-repeat` is transformed to `repeat-x`.
     if (m_xValue != m_yValue) {
         if (m_xValue == CSSValueRepeat && m_yValue == CSSValueNoRepeat)
-            return nameString(CSSValueRepeatX);
+            return maker(nameString(CSSValueRepeatX));
```
Comment (darinadler) [inline Source/WebCore/css/CSSBackgroundRepeatValue.cpp:52]: There’s a possible efficiency issue here. If we want a string, then `nameString` is perfect since the string is already allocated. If we just want to append then `nameLiteral` is slightly more efficient.

File: Source/WebCore/css/CSSBackgroundRepeatValue.cpp [hunk 3]
```diff
@@ -44,17 +44,27 @@ Ref<CSSBackgroundRepeatValue> CSSBackgroundRepeatValue::create(CSSValueID repeat
     return adoptRef(*new CSSBackgroundRepeatValue(repeatXValue, repeatYValue));
 }
 
-String CSSBackgroundRepeatValue::customCSSText() const
+template<typename Maker> decltype(auto) CSSBackgroundRepeatValue::serialize(Maker&& maker) const
```
Comment (darinadler) [inline Source/WebCore/css/CSSBackgroundRepeatValue.cpp:47]: How did you decide when to do this dual mode thing and when to just require `StringBuilder`?

File: Source/WebCore/css/CSSBasicShapes.cpp [hunk 4]
```diff
@@ -229,92 +234,115 @@ bool CSSXywhValue::equals(const CSSXywhValue& other) const
...
+
+    auto serializeRadiiAxis = [&](const auto (&r)[4]) {
+        if (!r[3]->equals(*r[1])) {
+            r[0]->cssText(builder);
+            builder.append(' ');
+            r[1]->cssText(builder);
+            builder.append(' ');
+            r[2]->cssText(builder);
+            builder.append(' ');
+            r[3]->cssText(builder);
```
Comment (darinadler) [inline Source/WebCore/css/CSSBasicShapes.cpp:309]: Code like this makes it clear that an interface built more on concatenation might lead to a cleaner idiom, and even possible more efficiency since we might resize a smaller number of times. But I don’t know how to do that without code bloat.

File: Source/WebCore/css/CSSCounterStyleDescriptors.cpp [hunk 5]
```diff
@@ -438,16 +470,72 @@ String CSSCounterStyleDescriptors::systemCSSText() const
...
+}
+
+// MARK: - Negative
+
+bool CSSCounterStyleDescriptors::NegativeSymbols::hasCSSText() const
+{
+    return true;
+}
+
+String CSSCounterStyleDescriptors::NegativeSymbols::cssText() const
```
Comment (darinadler) [inline Source/WebCore/css/CSSCounterStyleDescriptors.cpp:499]: It kind of seems like if the implementation is just using a `StringBuilder`, then it shouldn’t be a member function. Caller can use a helper to make the string in an idiom that is not too wordy, but having the function makes it seem like it’s more efficient than it really is? Maybe.

File: Source/WebCore/css/CSSCounterStyleDescriptors.cpp [hunk 6]
```diff
@@ -501,36 +650,131 @@ String CSSCounterStyleDescriptors::padCSSText() const
...
+// MARK: - Symbols
+
+bool CSSCounterStyleDescriptors::Symbol::hasCSSText() const
+{
+    if (isCustomIdent)
+        return !text.isEmpty();
+
+    // If the symbol is not a custom ident, it is serialized as a string, which
+    // will always at least include quotation marks, even for empty strings.
+    return true;
```
Comment (darinadler) [inline Source/WebCore/css/CSSCounterStyleDescriptors.cpp:697]: This comment is so clear, and given that, the code I would expect is:
```
return !(isCustomIdent && text.isEmpty());
```
I guess I’m a fan of boolean expressions when they can be readable.

**Review by darinadler — APPROVED** [review]
I have the following high level thoughts on this:
1) Would be nice to retain the property that we can reuse an existing string when the entire serialized string happens to be a CSS name. I think we did retain that in both code paths.
2) Would be nice to compute the size of the string and do a single allocation rather than doing the StringBuilder style resizing buffer in cases where it’s allowed. But coming up with the idiom that works that way while keeping the source code simple is probably not easy.
3) People will be writing more serialization code in the future, so we need to make it clear how to do this; the multiple different patterns might confuse them.
4) Although there is an economy, elegance, and even some clarity in using a single name, I am not sure that overloading with the same name is the best way to make pairs of functions where one has a return value and the other appends to an existing string builder. Naming scheme might help make this comprehensible.
5) This change should be a significant enough improvement that we should be able to measure the speedup from doing less memory allocation. I wonder what benchmark could be used to show this? There might even be a memory fragmentation savings, although that is less certain.
